cdc/kv/client_v2.go

```diff
@@ -0,0 +1,287 @@
+	case worker.inputCh <- &regionStatefulEvent{
+		changeEvent: event,
+		state:       state,
+	}:
+	}
+	return nil
+}
+
+func (s *eventFeedSession) sendResolvedTsV2(
+	ctx context.Context,
+	g *errgroup.Group,
+	resolvedTs *cdcpb.ResolvedTs,
+	worker *regionWorker,
+	addr string,
+) error {
+	for _, regionID := range resolvedTs.Regions {
+		state, ok := worker.getRegionState(regionID)
+		if ok {
+			if state.isStopped() {
+				log.Warn("drop resolved ts due to region feed stopped",
+					zap.Uint64("regionID", regionID),
+					zap.Uint64("requestID", state.requestID),
+					zap.String("addr", addr))
+				continue
+			}
+			select {
+			case worker.inputCh <- &regionStatefulEvent{
+				resolvedTs: resolvedTs,
+				state:      state,
+			}:
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+	}
+	return nil
+}
+
+// receiveFromStreamV2 receives gRPC messages from a stream continuously and sends
+// messages to region worker, if `stream.Recv` meets error, this routine will exit
+// silently. As for regions managed by this routine, there are two situations:
+// 1. established regions: a `nil` event will be sent to region worker, and region
+//    worker call `s.onRegionFail` to re-establish these regions.
+// 2. pending regions: call `s.onRegionFail` for each pending region before this
+//    routine exits to establish these regions.
+func (s *eventFeedSession) receiveFromStreamV2(
+	ctx context.Context,
+	g *errgroup.Group,
+	addr string,
+	storeID uint64,
+	stream cdcpb.ChangeData_EventFeedClient,
+	pendingRegions *syncRegionFeedStateMap,
+	limiter *rate.Limiter,
+) error {
+	// Cancel the pending regions if the stream failed. Otherwise it will remain unhandled in the pendingRegions list
+	// however not registered in the new reconnected stream.
+	defer func() {
+		log.Info("stream to store closed", zap.String("addr", addr), zap.Uint64("storeID", storeID))
+
+		remainingRegions := pendingRegions.takeAll()
+		for _, state := range remainingRegions {
+			err := s.onRegionFail(ctx, regionErrorInfo{
+				singleRegionInfo: state.sri,
+				err:              cerror.ErrPendingRegionCancel.GenWithStackByArgs(),
+			})
+			if err != nil {
+				// The only possible is that the ctx is cancelled. Simply return.
+				return
+			}
+		}
+		s.workersLock.Lock()
+		delete(s.workers, addr)
+		s.workersLock.Unlock()
+	}()
+
+	captureAddr := util.CaptureAddrFromCtx(ctx)
+	changefeedID := util.ChangefeedIDFromCtx(ctx)
+	metricSendEventBatchResolvedSize := batchResolvedEventSize.WithLabelValues(captureAddr, changefeedID)
+
+	s.workersLock.Lock()
+	worker, ok := s.workers[addr]
+	if !ok {
+		worker = &regionWorker{
+			session:        s,
+			limiter:        limiter,
+			inputCh:        make(chan *regionStatefulEvent, 1024),
+			outputCh:       s.eventCh,
+			regionStates:   make(map[uint64]*regionFeedState),
+			enableOldValue: s.enableOldValue,
+		}
+		s.workers[addr] = worker
+	}
+	s.workersLock.Unlock()
+
+	g.Go(func() error {
+		return worker.run(ctx)
```

Comment:
if `worker` is already exist in `s.workers`, should we still call the `worker.run` ?

Reply:
always create a new worker

Reply:
How many workers in total, let's say there are 1k tables and 10k regions?

Reply:
This is the first part of refactor, one stream, one worker.
Besides, We don't need too many workers for each stream